docs/book/src/developer/architecture/controllers/cluster.md

```diff
@@ -50,6 +50,9 @@ is a map, defined as `map[string]FailureDomainSpec`. A unique key must be used f
     - `controlPlane` (bool): indicates if failure domain is appropriate for running control plane instances.
     - `attributes` (`map[string]string`): arbitrary attributes for users to apply to a failure domain.
 
+Note: once any of `failureReason` or `failureMessage` surface on the cluster who is referencing the infrastructureCluster object, 
+they cannot be restored anymore (it is considered a terminal error).
```

Comment:
```suggestion
they cannot be restored anymore without user intervention (it is considered a terminal error).
```

cannot be restored is kind of hard may leading to the wrong conclusion, WDYT? (also tehn for the other occurencies)

Reply:
I think if that happens it should not be possible to restore them anymore. If someone sets failureReason / failureMessage in a case that is not terminal that is a bug in the controller who did it

Reply:
> That means that controllers of bootstrap, control-plane and infrastructure providers should not set failureReason/Message for something that can be fixed with user intervention (or in any other way).

I think this would be a pretty strong and with that definition no controller should ever set `failureReason` or `failureMessage`. Which also means `failureMessage` and `failureReason` should never be set.

With "user intervention" I'm also thinking of: "for cleanup I have to delete this object" which somewhat fixes it for the user because the failureReason or failureMessage is gone.

Or asked the other way, what should I as user do if I find objects having a `failureReason` or `failureMessage` set? (I think I miss some context here :-) )

Reply:
> what should I as user do if I find objects having a failureReason or failureMessage set? (I think I miss some context here :-) )

As far as I'm aware of the only option for a user facing a terminal error are:
- If the fatal error is on the cluster, delete and re-create a new one
- If the fatal error is on the machine, delete and re-create a new one

But at this point I'm not really sure where your comment are leading to. Would it work if I change 
(it is considered a terminal error)
into
(it is considered a terminal error; the only way to recover is to delete and recreate the cluster)

Reply:
Ah ok, got it. 

I think that I just had a wrong understanding of "terminal error" (because when I reached one in the past, it not really was terminal --> so it was a wrong usage from a controller). I think its fine to keep as is then 👍  Thanks for helping me to clarify it.

Reply:
We can dive a bit deeper later if you want. I think it comes down to there shouldn't be a terminal/fatal error that is not terminal through user interaction

Reply:
If user interaction could solve it, controllers shouldn't signal terminal. (Assuming they want users to manually interact with infra at all)

Reply:
Note: in the regular case with much these fields lead to immediate deletion of machines (+/- considering max unavailable)

Reply:
Not a good way to encourage manual user actions 😅